File: app/src/main/java/com/musicplayer/OpenMusic/ui/settings/SettingsFragment.java
package com.musicplayer.OpenMusic.ui.settings;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import com.musicplayer.OpenMusic.OpenMusicApp;
import com.musicplayer.OpenMusic.data.SongsData;
import com.musicplayer.OpenMusic.ui.dir_browser.DirBrowserActivity;
import com.musicplayer.OpenMusic.ui.sleeptime.SleepTimeActivity;
import com.musicplayer.musicplayer.BuildConfig;
import com.musicplayer.musicplayer.R;

import java.util.Objects;

public class SettingsFragment extends PreferenceFragmentCompat {
    private SongsData songsData;
    private Host hostCallBack;

    @Override
    public void onCreate(@Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        songsData = SongsData.getInstance(requireContext());
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preferences, rootKey);
        Preference libPathPreference = findPreference(OpenMusicApp.PREFS_KEY_LIBRARY_PATHS);
        Preference versions = findPreference(OpenMusicApp.PREFS_KEY_VERSION);
        //Preference logging = findPreference(OpenMusicApp.PREFS_KEY_LOGGING);
        Preference sleeptime = findPreference(OpenMusicApp.PREFS_KEY_SLEEPTIME);
        Preference menuswitch = findPreference(OpenMusicApp.PREFS_KEY_MENUSWITCH);
        ActivityResultLauncher<Intent> launcher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
            if (result.getResultCode() == Activity.RESULT_OK)
                hostCallBack.onLibraryDirsChanged();
        });
        libPathPreference.setOnPreferenceClickListener(preference -> {
            if (!songsData.isDoneLoading()) {
                Toast.makeText(requireContext(), R.string.settings_cannot_change_lib, Toast.LENGTH_SHORT).show();
                return false;
            }
            Intent intent = new Intent(getContext(), DirBrowserActivity.class);
            launcher.launch(intent);
            return true;
        });
        sleeptime.setOnPreferenceClickListener(preference -> {
            Intent intent = new Intent(getContext(), SleepTimeActivity.class);
            launcher.launch(intent);
            return true;
        });
        menuswitch.setOnPreferenceClickListener(preference -> {
            Objects.requireNonNull(getActivity()).recreate();
            return true;
        });

        versions.setSummary(getString(R.string.about_version, BuildConfig.VERSION_NAME));
    }

    /**
     * If the fragment is being attached to another activity
     *
     * @param context The context of the app
     */
    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        try {
            this.hostCallBack = (Host) context;
            // If implementation is missing
        } catch (final ClassCastException e) {
            throw new ClassCastException(context + " must implement SettingsFragment.Host");
        }
    }

    public interface Host {
        void onLibraryDirsChanged();
    }
}

File: app/src/main/kotlin/com/musicplayer/openmusic/ui/settings/SettingsFragment.kt
package com.musicplayer.openmusic.ui.settings

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.musicplayer.openmusic.BuildConfig
import com.musicplayer.openmusic.OpenMusicApp
import com.musicplayer.openmusic.R
import com.musicplayer.openmusic.data.SongsData
import com.musicplayer.openmusic.data.SongsData.Companion.getInstance
import com.musicplayer.openmusic.ui.dir_browser.DirBrowserActivity
import com.musicplayer.openmusic.ui.sleeptime.SleepTimeActivity

class SettingsFragment : PreferenceFragmentCompat() {
    private var songsData: SongsData? = null
    private var hostCallBack: Host? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        songsData = getInstance(requireContext())
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)
        val libPathPreference = findPreference<Preference>(OpenMusicApp.PREFS_KEY_LIBRARY_PATHS)
        val versions = findPreference<Preference>(OpenMusicApp.PREFS_KEY_VERSION)
        //Preference logging = findPreference(OpenMusicApp.PREFS_KEY_LOGGING);
        val sleeptime = findPreference<Preference>(OpenMusicApp.PREFS_KEY_SLEEPTIME)
        val menuswitch = findPreference<Preference>(OpenMusicApp.PREFS_KEY_MENUSWITCH)
        val launcher =
            registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult -> if (result.resultCode == Activity.RESULT_OK) hostCallBack!!.onLibraryDirsChanged() }
        assert(libPathPreference != null)
        libPathPreference!!.onPreferenceClickListener =
            Preference.OnPreferenceClickListener setOnPreferenceClickListener@{
                if (!songsData!!.isDoneLoading) {
                    Toast.makeText(
                        requireContext(),
                        R.string.settings_cannot_change_lib,
                        Toast.LENGTH_SHORT
                    ).show()
                    return@setOnPreferenceClickListener false
                }
                val intent = Intent(context, DirBrowserActivity::class.java)
                launcher.launch(intent)
                true
            }
        assert(sleeptime != null)
        sleeptime!!.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                val intent = Intent(context, SleepTimeActivity::class.java)
                launcher.launch(intent)
                true
            }
        assert(menuswitch != null)
        menuswitch!!.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                requireActivity().recreate()
                true
            }
        assert(versions != null)
        versions!!.summary = getString(R.string.about_version, BuildConfig.VERSION_NAME)
    }

    /**
     * If the fragment is being attached to another activity
     *
     * @param context The context of the app
     */
    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            hostCallBack = context as Host
            // If implementation is missing
        } catch (e: ClassCastException) {
            throw ClassCastException("$context must implement SettingsFragment.Host")
        }
    }

    interface Host {
        fun onLibraryDirsChanged()
    }
}